imbi/endpoints/project_ssm_configuration.py

```diff
@@ -134,6 +134,58 @@ async def get(self, *args, **kwargs) -> None:
         } for name, d in params_by_path.items()]
         self.send_response(output)
 
+    async def post(self, *args, **kwargs) -> None:
+        if not self.settings['google']['enabled']:
+            raise errors.Forbidden('Google integration is disabled',
+                                   title='Google integration is disabled')
+
+        if not self.settings['google']['integration_name']:
+            raise errors.Forbidden(
+                'No Google integration specified in configuration',
+                title='No Google integration specified in configuration')
+
+        tokens = await self.current_user.fetch_integration_tokens(
+            self.settings['google']['integration_name'])
+        if len(tokens) == 0:
+            raise errors.Forbidden(
+                'No OAuth 2.0 tokens for %s',
+                self.current_user.username,
+                title=('Not authorized to access SSM in this namespace &'
+                       ' environment'))
+        google_tokens = tokens[0]
+
+        aws_session = aioboto3.Session()
+        body = self.get_request_body()
+        for param in body['values']:
+            result = await self.postgres_execute(
+                self.GET_ROLE_ARN_SQL, {
+                    'environment': param['environment'],
+                    'namespace_id': body['namespace_id']
+                }, 'get-role-arn')
```

Comment:
Since this is for a single project, you should be able to move this output of the loop right? The ARN should always be the one from the namespace and a project can only be in one namespace. _The same should be true of the AWS credentials_ though they could expire mid-loop I guess.

Reply:
hmm... I guess that we will potentially have a separate ARN for each environment so maybe a per-environment cache.

Reply:
The loop is over each param in the POST body, which by how I've set up the frontend is only a max of one per environment, so the loop makes sense.

But thinking about it, we could run through the params once at first, collect all the environments, and then do a single query that fetches all the ARNs for those environments using `IN` and store environment -> ARN in a dict. Then loop through the params again and reference the ARNs as needed. This would make it a single SQL query and a little more efficient if we ever decide to add multiple params in the same environment at once.

Reply:
actually ... nevermind ... I was thrown by this being a list of per-environment values. What happens if the client passes multiple values for the same environment? My guess is that they will get a `400` with `ParameterAlreadyExists`.

What do you think about changing the API slightly to represent the values as a dict where the environment name is the key and the value is the parameter value? That would sidestep the same environment being the list more than once. It will look a little strange if we have spaces in the environment names but I don't think that it would cause any problems

Reply:
The endpoint itself is project-agnostic, unlike the GET endpoint. It just takes an array of params and tries to PUT them given the credentials of the user.

That's also got me thinking... now that the prefix template is in settings perhaps the GET should be project-agnostic as well. The frontend can send over the templated prefix and the environment list and it can totally skip all the project-specific code and query.

Reply:
That requires moving all this stuff out from under /projects, but the more I think about it the more I like it. It's more modular and single-responsibility.

Reply:
> What do you think about changing the API slightly to represent the values as a dict where the environment name is the key

I actually did this at first and thought that an array was more generic. It really depends on how project-oriented we want to make this.

Reply:
> The endpoint itself is project-agnostic

I was wondering if there was anything gating that other than it [having the project ID in the URL](https://github.com/AWeber-Imbi/imbi-api/blob/main/imbi/endpoints/__init__.py#L119-L121)

> It really depends on how project-oriented we want to make this

This is one of the rubs between an API and the UI. The UI is going to be in terms of a project because that is what makes sense there. We _could_ offer a non-project view but you should just use the AWS web UI for that. So, the UI is going to be project-centric but there isn't anything that ties our API to that. We could have the UI take the key prefix and generate a generic "update this set of SSM params" request to the API. I could definitely see the benefit of that at the API level. However, it will complicate the implementation since you will have to be very careful about caching tokens since we don't want to hit the database for what is most likely 1 token per environment (based on the UI).

It would also change the get to be "fetch me a bunch of SSM values" instead of the per-project version _if we go that direction_. I don't know .... most of the Imbi API is project-centric so it will feel a little out of place if it is not project-centric. Keeping it project-centric also let's us make _valid_ assumptions in the implementation with respect to the number of tokens involved -- it is a maximum of 1 per environment as opposed to 1 per SSM key in the pathological generic case.

I could see it going either way. I am leaning slightly towards the project-specific side of things since that is more consistent with the current API though I am breaking that a little more every day.

Reply:
Been thinking on this, I agree we should keep it project-oriented. I'm drawn to generic because it's composable, but from a design perspective everything is project-oriented and we should keep it that way.

I also hadn't originally considered that if we go full generic then we start need to extracting namespaces based on the prefix template, since namespace determines which credentials are used (vs project-oriented, where we get namespace bundled in the project for free). We'd need to leverage the SSM prefix template anyway but ignore the project part. I wouldn't be concerned about grabbing and caching environment-based tokens, but it's not just that, it's environment × namespace number of tokens.

To keep it project-oriented I do think I should change a couple things for POST:
- send an object with environment keys, instead of an array
- either validate the SSM paths adhere to the prefix for that project, or perhaps even better, have requests send over paths _after_ the prefix. The backend can generate the prefix itself (just like the GET endpoint does) and append them together. The frontend can have the prefix uneditable in front of the form field.